ui/pages/transaction/controller.go

```diff
@@ -14,18 +18,119 @@ func (m ViewModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 	return m.HandleMessage(msg)
 }
 
-// HandleMessage is called by the viewport to update it's Model
+func (m *ViewModel) UpdateTxnURLAndQRCode() error {
+	accountStatus := m.State.Accounts[m.Data.Address].Status
+
+	m.hint = ""
+
+	senderAddress, err := types.DecodeAddress(m.Data.Address)
+	if err != nil {
+		return err
+	}
+
+	var isOnline bool
+	switch accountStatus {
+	case "Online":
+		isOnline = true
+	case "Offline":
+		isOnline = false
+	case "NotParticipating": // This status means the account can never participate in consensus
+		m.urlTxn = ""
+		m.asciiQR = ""
+		m.hint = fmt.Sprintf("%s is NotParticipating. Cannot register key.", senderAddress)
+		return nil
+	}
+
+	// Construct Transaction
+	tx := types.Transaction{}
+
+	if !isOnline { // TX take account online
+		var stateProofPk types.MerkleVerifier
+		copy(stateProofPk[:], (*m.Data.Key.StateProofKey)[:])
+
+		tx = types.Transaction{
+			Type: types.KeyRegistrationTx,
+			Header: types.Header{
+				Sender: senderAddress,
+				Fee:    1000, //TODO: get proper fee
+			},
+			KeyregTxnFields: types.KeyregTxnFields{
+				VotePK:          types.VotePK(m.Data.Key.VoteParticipationKey),
+				SelectionPK:     types.VRFPK(m.Data.Key.SelectionParticipationKey),
+				StateProofPK:    types.MerkleVerifier(stateProofPk),
+				VoteFirst:       types.Round(m.Data.Key.VoteFirstValid),
+				VoteLast:        types.Round(m.Data.Key.VoteLastValid),
+				VoteKeyDilution: uint64(m.Data.Key.VoteKeyDilution),
+			},
+		}
+
+		// Update hint if no error
+		defer func() {
+			if err == nil {
+				m.hint = fmt.Sprintf("Scan this QR code to take %s Online.", senderAddress)
+			}
+		}()
+
+	} else { // TX to take account offline
+		tx = types.Transaction{
+			Type: types.KeyRegistrationTx,
+			Header: types.Header{
+				Sender: senderAddress,
+				Fee:    1000, //TODO: get proper fee
+			},
+		}
+
+		// Update hint if no error
+		defer func() {
+			if err == nil {
+				m.hint = fmt.Sprintf("Scan this QR code to take %s Offline.", senderAddress)
+			}
+		}()
+	}
+
+	// Construct QR Code
+	kr, err := encoder.MakeQRKeyRegRequest(
```

Comment:
This method calls the decoder and receives a transaction. We need to remove the transaction and sdk overhead here and just built the `AUrlTxn` with the objects already in memory.

```go
kr := &AUrlTxn{
  AUrlTxnKeyCommon: AUrlTxnKeyCommon{
	  Sender: m.Data.Address,
	  Type:   "keyreg",
  },
  AUrlTxnKeyreg: AUrlTxnKeyreg{
	  VotePK:          PtrOrNil(m.Data.Key.****), // Add data from Key
	  // .... Rest of the keys
  }
}
```